### PR TITLE
Compiled but untested iteration for Publisher modifications

### PR DESCRIPTION
**Pull request by kuznetsovmoci**
Some new code in publish, rmw_publish and rmw_create_publisher methods is temporary, until we create underlying type support logic.

### CODE REVIEW

File: rmw_opendds_cpp/include/rmw_opendds_cpp/opendds_static_publisher_info.hpp [hunk 1]
```diff
@@ -50,6 +50,36 @@ class OpenDDSPublisherListener : public DDS::PublisherListener
     return current_count_;
   }
 
+  void on_offered_deadline_missed(
+    ::DDS::DataWriter_ptr writer,
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/include/rmw_opendds_cpp/opendds_static_publisher_info.hpp]: Do we need to start from the global namespace? Can't it just be `DDS::DataWriter_ptr writer`

File: rmw_opendds_cpp/src/rmw_publisher.cpp [hunk 2]
```diff
@@ -110,20 +111,20 @@ rmw_create_publisher(
...
   DDS::PublisherQos publisher_qos;
   DDS::ReturnCode_t status;
-  DDS::Publisher * dds_publisher = nullptr;
-  DDS::DataWriter * topic_writer = nullptr;
-  DDS::Topic * topic = nullptr;
-  DDS::TopicDescription * topic_description = nullptr;
+  DDS::Publisher_var dds_publisher;
+  DDS::DataWriter_var topic_writer;
+  DDS::Topic_var topic;
+  DDS::TopicDescription_var topic_description;
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/rmw_publisher.cpp]: Can we not initialize these locals and the point of declaration?

File: rmw_opendds_cpp/src/rmw_publisher.cpp [hunk 3]
```diff
@@ -110,20 +111,20 @@ rmw_create_publisher(
...
   void * info_buf = nullptr;
   void * listener_buf = nullptr;
   OpenDDSPublisherListener * publisher_listener = nullptr;
   OpenDDSStaticPublisherInfo * publisher_info = nullptr;
   rmw_publisher_t * publisher = nullptr;
   std::string mangled_name = "";
+  OpenDDSStaticSerializedDataTypeSupport_var ts = new OpenDDSStaticSerializedDataTypeSupportImpl();
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/rmw_publisher.cpp:124]: Recognizing that we're focused on construction and not deconstruction at this point in the project, is the `delete` something we need to do or is this a smart pointer (or similar) type?
```suggestion
//TODO add delete
  OpenDDSStaticSerializedDataTypeSupport_var ts = new OpenDDSStaticSerializedDataTypeSupportImpl();
```

File: rmw_opendds_cpp/src/rmw_publisher.cpp [hunk 4]
```diff
@@ -134,21 +135,11 @@ rmw_create_publisher(
...
-  // which only publishes DDS_Octets
-  // The purpose of this is to send only raw data DDS_Octets over the wire,
-  // advertise the topic however with a type of the message, e.g. std_msgs::msg::dds_::String
-  //status = OpenDDSStaticSerializedDataSupport_register_external_type(
-  //  participant, type_name.c_str(), type_code);
-  if (status != DDS::RETCODE_OK) {
-    RMW_SET_ERROR_MSG("failed to register external type");
+  publisher->can_loan_messages = false;
+
+  // Register TypeSupport here ????????????????????????
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/rmw_publisher.cpp]:
```suggestion
  // TODO: Register TypeSupport here ????????????????????????
```

File: rmw_opendds_cpp/src/rmw_publisher.cpp [hunk 5]
```diff
@@ -274,14 +265,6 @@ rmw_create_publisher(
...
   node_info->publisher_listener->trigger_graph_guard_condition();
 
-// TODO(karsten1987): replace this block with logging macros
-#ifdef DISCOVERY_DEBUG_LOGGING
-  fprintf(stderr, "******* Creating Publisher Details: ********\n");
-  fprintf(stderr, "Publisher topic %s\n", topic_writer->get_topic()->get_name());
-  fprintf(stderr, "Publisher address %p\n", static_cast<void *>(dds_publisher));
-  fprintf(stderr, "******\n");
-#endif
-
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/rmw_publisher.cpp:284]: Do we want some logging at this level?

File: rmw_opendds_cpp/src/rmw_publisher.cpp [hunk 6]
```diff
@@ -481,25 +447,25 @@ rmw_destroy_publisher(rmw_node_t * node, rmw_publisher_t * publisher)
...
+    //  }
+    //  if (participant->delete_publisher(dds_publisher) != DDS::RETCODE_OK) {
+    //    RMW_SET_ERROR_MSG("failed to delete publisher");
+    //    return RMW_RET_ERROR;
+    //  }
+    //  publisher_info->dds_publisher_ = nullptr;
+    //} else if (publisher_info->topic_writer_) {
+    //  RMW_SET_ERROR_MSG("cannot delete datawriter because the publisher is null");
+    //  return RMW_RET_ERROR;
+    //}
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/rmw_publisher.cpp:470]: Is this code that isn't compiling but needs to be revisited or is it something we need to get rid of?

**Review by adamsj-ros — COMMENTED** [review]
Since this is in-work code, my biggest comment is making sure in code comments to state what needs to be done so Jiang and others know what to do.

File: rmw_opendds_cpp/src/rmw_publisher.cpp (hunk 3 above)
Comment (iguessthislldo) [reply rmw_opendds_cpp/src/rmw_publisher.cpp:124]: `_var` is a smart pointer.

File: rmw_opendds_cpp/include/rmw_opendds_cpp/opendds_static_publisher_info.hpp (hunk 1 above)
Comment (kuznetsovmoci) [reply rmw_opendds_cpp/include/rmw_opendds_cpp/opendds_static_publisher_info.hpp]: Fixed

File: rmw_opendds_cpp/src/rmw_publisher.cpp (hunk 4 above)
Comment (kuznetsovmoci) [reply rmw_opendds_cpp/src/rmw_publisher.cpp]: Done

File: rmw_opendds_cpp/src/rmw_publisher.cpp (hunk 5 above)
Comment (kuznetsovmoci) [reply rmw_opendds_cpp/src/rmw_publisher.cpp:284]: Added 'TODO' comment

File: rmw_opendds_cpp/src/rmw_publisher.cpp (hunk 6 above)
Comment (kuznetsovmoci) [reply rmw_opendds_cpp/src/rmw_publisher.cpp:470]: This code is not needed after pointers have been replaced with smart pointers.

File: rmw_opendds_cpp/src/rmw_publisher.cpp (hunk 2 above)
Comment (kuznetsovmoci) [reply rmw_opendds_cpp/src/rmw_publisher.cpp]: Done

**Pull request merged**